op-service/eth/transactions_test.go

```diff
@@ -30,111 +30,105 @@ func (m *MockL1Client) HeaderByNumber(ctx context.Context, number *big.Int) (*ty
 
 func TestTransactions_checkRecentTxs(t *testing.T) {
 	tests := []struct {
-		name               string
-		currentBlock       uint64
-		blockConfirms      uint64
-		previousNonceBlock uint64
-		expectedBlockNum   uint64
-		expectedFound      bool
+		name             string
+		currentBlock     uint64
+		blockConfirms    uint64
+		expectedBlockNum uint64
+		expectedFound    bool
+		nonces           map[uint64]uint64 // maps blockNum --> nonce
 	}{
 		{
-			// Blocks       495 496 497 498 499 500
-			// Nonce          5   5   5   6   6   6
-			// call NonceAt   x   -   x   x   x   x
-			name:               "NonceDiff_3Blocks",
-			currentBlock:       500,
-			blockConfirms:      5,
-			previousNonceBlock: 497,
-			expectedBlockNum:   498,
-			expectedFound:      true,
+			name:             "nonceDiff_LowerBound",
+			currentBlock:     500,
+			blockConfirms:    5,
+			expectedBlockNum: 496,
+			expectedFound:    true,
+			nonces: map[uint64]uint64{
+				495: 5,
+				496: 6,
+				497: 6,
+				498: 6,
+				499: 6,
+				500: 6,
+			},
 		},
 		{
-			// Blocks       495 496 497 498 499 500
-			// Nonce          5   5   5   5   5   6
-			// call NonceAt   x   -   -   -   x   x
-			name:               "NonceDiff_1Block",
-			currentBlock:       500,
-			blockConfirms:      5,
-			previousNonceBlock: 499,
-			expectedBlockNum:   500,
-			expectedFound:      true,
+			name:             "nonceDiff_midRange",
+			currentBlock:     500,
+			blockConfirms:    5,
+			expectedBlockNum: 497,
+			expectedFound:    true,
+			nonces: map[uint64]uint64{
+				495: 5,
+				496: 5,
+				497: 6,
+				498: 6,
+				499: 6,
+				500: 6,
+			},
 		},
 		{
-			// Blocks       495 496 497 498 499 500
-			// Nonce          6   6   6   6   6   6
-			// call NonceAt   x   -   -   -   -   x
-			name:               "NonceUnchanged",
-			currentBlock:       500,
-			blockConfirms:      5,
-			previousNonceBlock: 400,
-			expectedBlockNum:   495,
-			expectedFound:      false,
+			name:             "nonceDiff_upperBound",
+			currentBlock:     500,
+			blockConfirms:    5,
+			expectedBlockNum: 500,
+			expectedFound:    true,
+			nonces: map[uint64]uint64{
+				495: 5,
+				496: 5,
+				497: 5,
+				498: 5,
+				499: 5,
+				500: 6,
+			},
+		},
+		{
+			name:             "nonce_unchanged",
+			currentBlock:     500,
+			blockConfirms:    5,
+			expectedBlockNum: 495,
+			expectedFound:    false,
+			nonces: map[uint64]uint64{
+				495: 6,
+				496: 6,
+				497: 6,
+				498: 6,
+				499: 6,
+				500: 6,
+			},
+		},
+		{
+			name:             "reorg",
+			currentBlock:     500,
+			blockConfirms:    5,
+			expectedBlockNum: 495,
+			expectedFound:    false,
+			nonces: map[uint64]uint64{
+				495: 5,
+				496: 7,
+				497: 6,
+				498: 6,
+				499: 6,
+				500: 6,
+			},
```

Comment:
If we implement reorg behavior as a restart of the search, this test case becomes a bit more complex to set up, as we'd need to return different nonces on successive calls.

Reply:
https://github.com/ethereum-optimism/optimism/pull/11232/commits/9f6009764310d8a35f91b2cbe9fa5281f8616eb2